### PR TITLE
[BEAM-7637] Migration s3 on sdkv2

### PR DESCRIPTION
**Pull request by RaphaelSanamyan**
*This PR is adding S3FileSystem v2 to the amazon-web-services2 submodule, and which uses AWS SDK V2. There are a few changes comparing to S3FileSystem V1:

1) Adding SSECustomerKey class since their own class was removed from SDKV2.
2) Adding tests for SSECustomerKey.
3) Removing all deprecated methods from tests. 
3) For SSE-C encryption, SSECustomerKey and SSECustomerAlgorithm must be specified.
4) For SSE-KMS encryption, SSEAlgorithm and SSEKMSKeyId must be specified.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by RaphaelSanamyan** [comment]
@iemejia Could you please review this PR or suggest somebody who could do it.

**Comment by iemejia** [comment]
Hi sorry, given the size of the PR + a bit of lower availability these days please be patient. I will take a look as soon as I can.

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/S3ResourceId.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+  /** Matches a glob containing a wildcard, capturing the portion before the first wildcard. */
+  private static final Pattern GLOB_PREFIX = Pattern.compile("(?<PREFIX>[^\\[*?]*)[\\[*?].*");
+
+  private final String bucket;
+  private final String key;
+  private final @Nullable Long size;
+  private final @Nullable Date lastModified;
+
+  public static void main(String[] args) {
```
Comment (ilya-kozyrev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/S3ResourceId.java:52]: I think it not a necessary method

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/SSECustomerKey.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+package org.apache.beam.sdk.io.aws2.s3;
+
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class SSECustomerKey {
```
Comment (ilya-kozyrev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/SSECustomerKey.java:27]: Could you please add doc for class?

**Review by ilya-kozyrev — CHANGES_REQUESTED** [review]

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/S3ResourceId.java (hunk 1 above)
Comment (RaphaelSanamyan) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/S3ResourceId.java]: removed

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/SSECustomerKey.java (hunk 2 above)
Comment (RaphaelSanamyan) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/SSECustomerKey.java:28]: Add doc "/** Customer provided key for use with Amazon S3 server-side encryption. */" to class

**Comment by RaphaelSanamyan** [comment]
Run Python PreCommit

**Comment by RaphaelSanamyan** [comment]
run codecov/project/python

**Comment by RaphaelSanamyan** [comment]
retest this please

**Comment by RaphaelSanamyan** [comment]
Run Python PreCommit

**Review by ramazan-yapparov — COMMENTED** [review]
Small comments, otherwise LGTM

**Review by ilya-kozyrev — APPROVED** [review]

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/S3FileSystem.java [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+    checkState(
+        !nonGlobMatches.hasNext(),
+        "Internal error encountered in S3Filesystem: expected no more elements in nonGlobMatches.");
+
+    return matchResults.build();
+  }
+
+  /** Gets {@link MatchResult} representing all objects that match wildcard-containing paths. */
+  @VisibleForTesting
+  List<MatchResult> matchGlobPaths(Collection<S3ResourceId> globPaths) throws IOException {
```
[the remark this answers is not in the document]
Comment (R1bell) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/S3FileSystem.java:194]: I don't mind and already rewrote it.

[the remark this answers is not in the document]
Comment (RaphaelSanamyan) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/S3FileSystem.java:194]: I don't mind and already rewrote it.

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/options/S3Options.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+   * Provide the default s3 upload buffer size in bytes: 64MB if more than 512MB in RAM are
+   * available and 5MB otherwise.
+   */
+  class S3UploadBufferSizeBytesFactory implements DefaultValueFactory<Integer> {
+
+    public static final int MINIMUM_UPLOAD_BUFFER_SIZE_BYTES = 5_242_880;
+
+    @Override
+    public Integer create(PipelineOptions options) {
+      return Runtime.getRuntime().maxMemory() < 536_870_912
```
Comment (daria-malkova) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/options/S3Options.java:90]: Maybe give a name to this numbers? Move to constants?

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/s3/S3TestUtils.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    S3Options options = PipelineOptionsFactory.as(S3Options.class);
+    options.setAwsRegion("us-west-1");
+    options.setS3UploadBufferSizeBytes(5_242_880);
+    return options;
+  }
+
+  static S3Options s3OptionsWithPathStyleAccessEnabled() {
+    S3Options options = PipelineOptionsFactory.as(S3Options.class);
+    options.setAwsRegion("us-west-1");
+    options.setS3UploadBufferSizeBytes(5_242_880);
```
Comment (daria-malkova) [inline sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/s3/S3TestUtils.java:45]: Move this number to constants?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/S3ReadableSeekableByteChannel.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+import software.amazon.awssdk.core.exception.SdkServiceException;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.services.s3.model.HeadObjectRequest;
+
+@SuppressWarnings({
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+class S3ReadableSeekableByteChannel implements SeekableByteChannel {
```
Comment (daria-malkova) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/S3ReadableSeekableByteChannel.java:44]: Maybe give some docstrings to this class and classes below?

**Review by daria-malkova — APPROVED** [review]

**Review by ramazan-yapparov — APPROVED** [review]

**Review by aromanenko-dev — CHANGES_REQUESTED** [review]
Thank you for contribution!
I have several questions and notes.
1) Does this PR mostly replicate SDK v1 module and copies its codebase in a way to use S3 with SDK v2 + SSE changes?
2) Please, squash all commits into one or several atomic commits and avoid merging master or other branches (use `rebase` instead).
3) Please, update `CHANGES.md`

**Comment by RaphaelSanamyan** [comment]
@aromanenko-dev 
Hello, thank you for your comments.

1. Yes, the main codebase was taken from the first version of the S3Filesystem. Only the interaction with AWS S3 and SSE was changed.
2. Squashed to two commits. Thx for the notes.
3. CHANGES.md was updated.

**Comment by RaphaelSanamyan** [comment]
Run Java PreCommit

**Comment by RaphaelSanamyan** [comment]
Run Java PreCommit

**Comment by RaphaelSanamyan** [comment]
Run Java PreCommit

**Comment by RaphaelSanamyan** [comment]
Run Java PreCommit

**Review by aromanenko-dev — APPROVED** [review]
LGTM, thanks!

**Comment by aromanenko-dev** [comment]
Failed check are not related to this PR, see a discussion on [dev@](https://lists.apache.org/thread.html/re0a0e5a0d364f74a78d25f3abd6484c1f464ce59cb57af60e37caf34%40%3Cdev.beam.apache.org%3E)

**Pull request merged**